File: 2tdsr/enterprise-application-development/01-revisao/src/br/com/fiap/exercicio/bean/ContaCorrente.java
package br.com.fiap.exercicio.bean;

import java.util.Calendar;

import br.com.fiap.exercicio.exception.SaldoInsuficienteException;

public class ContaCorrente extends Conta {

	private TipoConta tipo;

	@Override
	public void retirar(Double valor) throws SaldoInsuficienteException {
		if (this.tipo == TipoConta.COMUM && super.saldo - valor < 0) {
			throw new SaldoInsuficienteException("Saldo insuficiente");
		}
		saldo -= valor;
	}

	public ContaCorrente() {
	}

	public ContaCorrente(Integer agencia, Integer numero, Calendar dataAbertura, Double saldo, TipoConta tipo) {
		super(agencia, numero, dataAbertura, saldo);
		this.tipo = tipo;
	}

	public TipoConta getTipoConta() {
		return tipo;
	}

	public void setTipoConta(TipoConta tipoConta) {
		this.tipo = tipoConta;
	}

}

File: 2tdsr/enterprise-application-development-asp-net/Fiap.Aula01/Fiap.Aula01.Exercicio01/Models/ContaCorrente.cs
using Fiap.Aula01.Exercicio01.Models;
using Fiap.Aula01.Exercicio01.Exceptions;

namespace Fiap.Aula01.Exercicio01.Models
{
    public class ContaCorrente : Conta
    {
        public TipoConta Tipo { get; set; }
        public decimal Limite { get; set; }

        public ContaCorrente(TipoConta tipo, int agencia, int numero, 
            IList<Cliente> clientes) : base(agencia, numero, clientes)
        {   
            Tipo = tipo;
            Limite = Tipo === TipoConta.Especial ? 500 : Tipo === TipoConta.Premium ? 1000 : 0;
            // switch (Tipo)
            // {
            //     case TipoConta.Comum: 
            //         Limite = 0;
            //         break;
            //     case TipoConta.Especial: 
            //         Limite = 500;
            //         break;
            //     case TipoConta.Premium: 
            //         Limite = 100;
            //         break;
            // }
            
        }
        
        public override void Retirar(decimal valor)
        {
            if(valor > Saldo + limite)
            {
                throw new SaldoInsuficienteException("Saldo insuficiente");
            }
            Saldo -= valor;
        }
    }

    enum TipoConta
    {
        Comum, 
        Especial,
        Premium
    }
}